ml-agents/mlagents/trainers/ghost/trainer.py

```diff
@@ -0,0 +1,253 @@
+# # Unity ML-Agents Toolkit
+# ## ML-Agent Learning (Ghost Trainer)
+
+# import logging
+from typing import Dict, List, Any, cast
+
+import numpy as np
+import logging
+
+from mlagents.trainers.brain import BrainParameters
+from mlagents.trainers.policy import Policy
+from mlagents.trainers.tf_policy import TFPolicy
+
+from mlagents.trainers.trainer import Trainer
+from mlagents.trainers.trajectory import Trajectory
+from mlagents.trainers.agent_processor import AgentManagerQueue
+
+LOGGER = logging.getLogger("mlagents.trainers")
+
+
+class GhostTrainer(Trainer):
+    def __init__(
+        self, trainer, brain_name, reward_buff_cap, trainer_parameters, training, run_id
+    ):
+        """
+        Responsible for collecting experiences and training trainer model via self_play.
+        :param trainer: The trainer of the policy/policies being trained with self_play
+        :param brain_name: The name of the brain associated with trainer config
+        :param reward_buff_cap: Max reward history to track in the reward buffer
+        :param trainer_parameters: The parameters for the trainer (dictionary).
+        :param training: Whether the trainer is set for training.
+        :param run_id: The identifier of the current run
+        """
+
+        super(GhostTrainer, self).__init__(
+            brain_name, trainer_parameters, training, run_id, reward_buff_cap
+        )
+
+        self.trainer = trainer
+
+        self.internal_policy_queues: List[AgentManagerQueue[Policy]] = []
+        self.learning_policy_queues: Dict[str, AgentManagerQueue[Policy]] = {}
+        self.internal_trajectory_queues: List[AgentManagerQueue[Trajectory]] = []
+
+        # assign ghost's stats collection to wrapped trainer's
+        self.stats_reporter = self.trainer.stats_reporter
+
+        self_play_parameters = trainer_parameters["self_play"]
+        self.window = self_play_parameters["window"]
```

Comment:
Default values for these? E.g. `self.window = self_play_parameters.get("window", 10)`

Otherwise, you need some nicer error handling for when these aren't set.

Reply:
Am generally in support of default values over throwing an error

Reply:
Fixed in the next push.